python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll

```diff
@@ -0,0 +1,172 @@
+  }
+
+  /** Gets a reference to a `SendGridAPIClient` instance call with `send` or `post`. */
+  private DataFlow::Node sendgridApiSendCall() {
+    result = sendgridApiClient() and
+    result.(DataFlow::AttrRead).getAttributeName() in ["send", "post"]
+  }
+
+  private DataFlow::Node sendgridContent(DataFlow::CallCfgNode contentCall, string mime) {
+    exists(StrConst mimeNode |
+      mimeNode.getText() = mime and
+      DataFlow::exprNode(mimeNode).(DataFlow::LocalSourceNode).flowsTo(contentCall.getArg(0)) and
+      result = contentCall.getArg(1)
+    )
+  }
+
+  private DataFlow::Node sendgridWrite(string attributeName) {
+    exists(DataFlow::AttrWrite attrWrite |
+      attrWrite.getObject().getALocalSource() = sendgridMailCall() and
+      attrWrite.getAttributeName() = attributeName and
+      result = attrWrite.getValue()
+    )
+  }
+
+  /**
+   * Gets a reference to `sg.send()` and `sg.client.mail.send.post()`.
+   *
+   * Given the following example:
+   *
+   * ```py
+   *  from_email = Email("from@example.com")
+   *  to_email = To("to@example.com")
+   *  subject = "Sending with SendGrid is Fun"
+   *  content = Content("text/html", request.args["html_content"])
+   *
+   *  mail = Mail(from_email, to_email, subject, content)
+   *
+   *  sg = SendGridAPIClient(api_key='SENDGRID_API_KEY')
+   *  response = sg.client.mail.send.post(request_body=mail.get())
+   *  ```
+   *
+   * * `this` would be `sg.client.mail.send.post(request_body=mail.get())`.
+   * * `getPlainTextBody()`'s result would be `none()`.
+   * * `getHtmlBody()`'s result would be `request.args["html_content"]`.
+   * * `getTo()`'s result would be `"to@example.com"`.
+   * * `getFrom()`'s result would be `"from@example.com"`.
+   * * `getSubject()`'s result would be `"Sending with SendGrid is Fun"`.
+   */
+  private class SendGridMail extends DataFlow::CallCfgNode, EmailSender::Range {
+    SendGridMail() { this.getFunction() = sendgridApiSendCall() }
+
+    override DataFlow::Node getPlainTextBody() {
+      result in [
+          sendgridMailCall().getArg(3), sendgridMailCall().getArgByName("plain_text_content")
+        ]
+      or
+      result in [
+          sendgridContent(sendgridMailHelper().getMember("Content").getACall(), "text/plain"),
+          sendgridContent(sendgridMailInstance().getMember("add_content").getACall(), "text/plain")
+        ]
+      or
+      result = sendgridWrite("plain_text_content")
+    }
+
+    override DataFlow::Node getHtmlBody() {
+      result in [sendgridMailCall().getArg(4), sendgridMailCall().getArgByName("html_content")]
+      or
+      result = sendgridMailInstance().getMember("set_html").getACall().getArg(0)
+      or
+      result in [
+          sendgridContent(sendgridMailHelper().getMember("Content").getACall(),
+            ["text/html", "text/x-amp-html"]),
+          sendgridContent(sendgridMailInstance().getMember("add_content").getACall(),
+            ["text/html", "text/x-amp-html"])
+        ]
```

Comment:
I like this construction, but again you have to synchronise on the instance. The call that `this` refer to may currently not be affected at all by the call returned by `sendgridMailHelper().getMember("Content").getACall()`.